LayoutTests/imported/w3c/web-platform-tests/pointerlock/pointerlock_unadjustedMovement.html

```diff
@@ -0,0 +1,42 @@
+<!DOCTYPE html>
+<html>
+<body>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="/resources/testdriver.js"></script>
+<script src="/resources/testdriver-actions.js"></script>
+<script src="/resources/testdriver-vendor.js"></script>
+</head>
+<body>
+    <h2>Description</h2>
+    <p>This test validates that pointer lock accepts the unadjustedMovement option.</p>
+    <hr/>
+
+    <button id="Button">lockTarget</button>
+    <div id="target">Target</div>
+
+    <script type="text/javascript" >
+        let button = document.getElementById('Button');
+        let target = document.getElementById('target');
+
+        async_test(t => {
+            button.addEventListener('mousedown', t.step_func(() => {
+                const p = target.requestPointerLock({unadjustedMovement: true});
+                p.then(t.step_func(() => {
+                    assert_equals(document.pointerLockElement, target);
+                    t.done();
+                })).catch(t.step_func((error) => {
+                    // requestPointerLock may throw NotSupportedError to say unadjustedMovement isn't supported.
+                    assert_throws_dom("NotSupportedError", () => { throw error; });
+                    // but to pass this test fully, it unadjustedMovement must be supported.
+                    assert_unreached("unadjustedMovement must be implemented.");
+                    t.done();
+                }));
+            }));
+        });
+
+        // Automated testing
+        test_driver.click(button);
```

Comment:
This could be simplified to?  

```suggestion
        async_test(async t => {
            await new Promise((resolve) => button.addEventListener('mousedown', resolve));
            try {
                await target.requestPointerLock({ unadjustedMovement: true });
                assert_equals(document.pointerLockElement, target);
            } catch (error) {
                // requestPointerLock may throw NotSupportedError to say unadjustedMovement isn't supported.
                assert_throws_dom("NotSupportedError", () => { throw error; });
            }
        });
        
        // Automated testing
        test_driver.click(button);
```

Reply:
I couldn't get that to work. A few things I ran into with that:

1. async_test => promise_test.
2. test_driver.click(button) runs before button.addEventListener.
3. So move test_driver.click into test function, but then we still don't have user activation.

```javascript
promise_test(async t => {
    const buttonClicked = new Promise((resolve) => button.addEventListener('mousedown', resolve));
    
    test_driver.click(button);
    await buttonClicked; // <-- no user activation here :(
    
    try {
        await target.requestPointerLock({ unadjustedMovement: true });
        assert_equals(document.pointerLockElement, target);
    } catch (error) {
        // requestPointerLock may throw NotSupportedError to say unadjustedMovement isn't supported.
        assert_throws_dom("NotSupportedError", () => { throw error; });
    }
});
```

```
FAIL pointerlock_unadjustedMovement assert_throws_dom: function "() => { throw error; }" threw object "NotAllowedError: Pointer lock requires a user gesture." that is not a DOMException NotSupportedError: property "code" is equal to 0, expected 9
```

Reply:
Sorry, I should have tested it... 

Can you try this (which I didn't test, but yolo 🙈):

```JS
promise_test(async t => {
    await test_driver.bless();
    assert_true(navigator.userActivation.isActive, "has user activation");
    try {
        await target.requestPointerLock({ unadjustedMovement: true });
        assert_equals(document.pointerLockElement, target);
    } catch (error) {
        // requestPointerLock may throw NotSupportedError to say unadjustedMovement isn't supported.
        assert_throws_dom("NotSupportedError", () => { throw error; });
    }
});
```

Reply:
Still no luck:

```
FAIL pointerlock_unadjustedMovement assert_throws_dom: function "() => { throw error; }" threw object "NotAllowedError: Pointer lock requires a user gesture." that is not a DOMException NotSupportedError: property "code" is equal to 0, expected 9
```

Though it is mighty suspicious that `navigator.userActivation.isActive` is true, and yet we're still getting NotAllowedError. I'll try to look into that further.

Reply:
ok yeah... that's is weird.

Reply:
Here's why:

`PointerLockController::requestPointerLock` tests `UserGestureIndicator::processingUserGesture()`, which is true only in the event handler for a user gesture.

`navigator.userActivation.isActive` tests `LocalDOMWindow::hasTransientActivation`, which is a more lenient check that activation happened within some time interval.

With that in mind, it's probably possible to still use promise_test but make sure that the requestPointerLock happens within the click event handler, but I am not sure if it will lead to a more clear or concise test.

Reply:
Ok, can we move away from use `UserGestureIndicator::processingUserGesture()`?... this should be using transient activation instead. 

> Pointer lock is a [transient activation-gated API](https://html.spec.whatwg.org/multipage/interaction.html#transient-activation-gated-api), therefore a [requestPointerLock](https://w3c.github.io/pointerlock/#dom-element-requestpointerlock)() call MUST fail if the [relevant global object](https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-global) of [this](https://webidl.spec.whatwg.org/#this) does not have [transient activation](https://html.spec.whatwg.org/multipage/interaction.html#transient-activation). This prevents locking upon initial navigation or re-acquiring lock without user's attention.

Reply:
(maybe we should do that as a follow-up... but this it's potentially going to be an interop issue if other browsers are using transient activation)

Reply:
Yeah, I think that needs to be a followup, depending on what the answers to [your questions](https://github.com/w3c/pointerlock/pull/49/files#r1373679855) about transient activation vs the engagement gesture defined in the spec change PR are.

Reply:
One wrinkle about moving to transient activation is this part of the spec:

> A [requestPointerLock](https://www.w3.org/TR/pointerlock-2/#dom-element-requestpointerlock)() call immediately after the [default unlock gesture](https://www.w3.org/TR/pointerlock-2/#dfn-default-unlock-gesture) MUST fail even when [transient activation](https://html.spec.whatwg.org/multipage/interaction.html#transient-activation) is available, to prevent malicious sites from acquiring an unescapable locked state through repeated lock attempts.

With that in mind, it's not as easy as just replacing `processingUserGesture()` with `hasTransientActivation()`.

I guess one solution might be by using `LocalDOMWindow::consumeTransientActivation()`?

Reply:
oh Lordy... ok, yes. It's going to need a mix of things.